kore/src/Kore/Reachability/Claim.hs

```diff
@@ -655,6 +669,180 @@ checkImplicationWorker (ClaimPattern.refreshExistentials -> claimPattern) =
+checkSimpleImplication left right existentials =
+    do
+        checkAssumptions
+        let definedConfig =
+                Pattern.andCondition left $
+                    from $ makeCeilPredicate leftTerm
+        trivial <-
+            fmap isBottom $
+                (liftSimplifier . SMT.Evaluator.filterMultiOr)
+                    =<< Pattern.simplify definedConfig
+
+        if trivial
+            then pure $ Implied (claimPattern, Just mempty)
+            else do
+                -- attempt term unification (to remember the substitution
+                unified <-
+                    Logic.observeAllT $
+                        evalEnvUnifierT
+                            Not.notSimplifier
+                            (unificationProcedure SideCondition.top leftTerm rightTerm)
+
+                -- for each unification result, attempt to refute the formula
+                remainders ::
+                    [(OrPattern RewritingVariableName, Substitution RewritingVariableName)] <-
+                    mapM (lift . checkUnifiedConsequent definedConfig) unified
+
+                let (successes, stucks) = partition (isBottom . fst) remainders
+
+                case (successes, stucks) of
+                    ((_bottom, subst) : _, _) ->
+                        -- success for at least one, return (ignore other failures)
+                        pure $ Implied (claimPattern, Just subst)
+                    ([], []) ->
+                        -- no successful unification, thus not implied
+                        pure $ NotImplied (claimPattern, Nothing)
+                    ([], stuck : _) ->
+                        -- successful unification but all stuck,
+                        -- return with first stuck term as antecedent
+                        let stuckClaim =
+                                (claimPattern :: ClaimPattern){left = OrPattern.toPattern sort (fst stuck)}
+                         in pure $ NotImpliedStuck (stuckClaim, Just $ snd stuck)
+  where
+    checkAssumptions = do
+        let showTerm = Pretty.renderString . Pretty.layoutOneLine . Pretty.pretty
+        -- must be function-like
+        withContext (showTerm leftTerm) $
+            koreFailWhen
+                (not $ isFunctionPattern leftTerm)
+                "The check implication step expects the antecedent term to be function-like."
+        -- RHS existentials must not capture free variables of LHS
+        let nameCollisions =
+                existentials `intersect` getFreeElementVariables (freeVariables left)
+        withContext ("LHS: " <> showTerm leftTerm)
+            . withContext ("RHS: " <> showTerm rightTerm)
+            . withContext ("existentials: " <> show (map unparse2 existentials))
+            $ koreFailWhen (not $ null nameCollisions) $
+                unwords
+                    ( "Existentials capture free variables of the LHS:" :
+                      map (show . unparse2) nameCollisions
+                    )
```

Comment:
I remember you had a question about this yesterday. I think this approach is good, they can take care of making sure there aren't collisions instead of delegating this to us. We'll see in the future if requirements change.